packages/react/src/components/cookieConsent/CookieConsentContext.tsx

```diff
@@ -0,0 +1,245 @@
+    ui: UiTexts;
+    tableHeadings: TableData;
+  };
+  language: {
+    languageOptions: { code: string; label: string }[];
+    current: string;
+    languageSelectorAriaLabel: string;
+    onLanguageChange: (newLanguage: string) => void;
+  };
+  onAllConsentsGiven?: (consents: ConsentObject) => void;
+  onConsentsParsed?: (consents: ConsentObject, hasUserHandledAllConsents: boolean) => void;
+};
+
+export type CookieConsentContextType = {
+  update: ConsentController['update'];
+  approveRequired: ConsentController['approveRequired'];
+  approveAll: ConsentController['approveAll'];
+  hasUserHandledAllConsents: () => boolean;
+  content: Content;
+  onAction: CookieConsentActionListener;
+  countApprovedOptional: () => number;
+  areGroupConsentsApproved: (consents: ConsentData[]) => boolean;
+};
+
+type CookieConsentContextProps = {
+  cookieDomain?: string;
+  children: React.ReactNode | React.ReactNode[] | null;
+  content: Content;
+};
+
+export const CookieConsentContext = createContext<CookieConsentContextType>({
+  update: () => undefined,
+  approveRequired: () => undefined,
+  approveAll: () => undefined,
+  hasUserHandledAllConsents: () => false,
+  content: {} as Content,
+  onAction: () => undefined,
+  countApprovedOptional: () => 0,
+  areGroupConsentsApproved: () => false,
+});
+
+export const getConsentsFromConsentGroup = (groups: ConsentGroup[]): ConsentList => {
+  return groups.reduce((ids, currentGroup) => {
+    currentGroup.consents.forEach((consentData) => {
+      ids.push(consentData.id);
+    });
+    return ids;
+  }, []);
+};
+
+export const Provider = ({ cookieDomain, children, content }: CookieConsentContextProps): React.ReactElement => {
+  const requiredConsents = content.requiredConsents
+    ? getConsentsFromConsentGroup(content.requiredConsents.groupList)
+    : undefined;
+  const optionalConsents = content.optionalConsents
+    ? getConsentsFromConsentGroup(content.optionalConsents.groupList)
+    : undefined;
+  const consentController = useMemo(() => create({ requiredConsents, optionalConsents, cookieDomain }), []);
+  const hasUserHandledAllConsents = () =>
+    consentController.getRequiredWithoutConsent().length === 0 && consentController.getUnhandledConsents().length === 0;
+
+  const mergeConsents = () => ({
+    ...consentController.getRequired(),
+    ...consentController.getOptional(),
+  });
+
+  const [, forceUpdate] = useState<number>(0);
+  const reRender = () => {
+    forceUpdate((p) => p + 1);
+  };
+
+  const save = () => {
+    const savedData = consentController.save();
+    if (hasUserHandledAllConsents()) {
+      reRender();
+      if (content.onAllConsentsGiven) {
+        content.onAllConsentsGiven(mergeConsents());
+      }
+    }
+    return savedData;
+  };
+
+  const update: CookieConsentContextType['update'] = (key, value) => {
+    consentController.update(key, value);
+  };
+  const approveSelectedAndRequired = () => {
+    consentController.approveRequired();
```

Comment:
Would `consentController.approveSelectedAndRequired();` describe the meaning of the function better?

Reply:
No, because is does not do that. consentController.approveRequired() sets all required consents "true".

Selected consents are set when clicked, but those are saved here  too.

The code has changes after this and naming has improved (IMHO)